go/vm/geth/processor.go

```diff
@@ -0,0 +1,645 @@
+	gasPrice := tx.GasPrice.ToUint256()
+	mgval := uint256.NewInt(uint64(tx.GasLimit))
+	mgval = mgval.Mul(mgval, gasPrice)
+	// Note: Opera doesn't need to check against gasFeeCap instead of gasPrice, as it's too aggressive in the asynchronous environment
+	balance := state.GetBalance(tx.Sender)
+	if have, want := balance.ToUint256(), mgval; have.Cmp(want) < 0 {
+		//skippedTxsNoBalanceMeter.Mark(1)
+		return fmt.Errorf("%w: address %v have %v want %v", errInsufficientFunds, tx.Sender, have, want)
+	}
+	// TODO: track block-wide gas usage
+	/*
+		if err := st.gp.SubGas(st.msg.Gas()); err != nil {
+			return err
+		}
+	*/
+	/*
+		st.gas += st.msg.Gas()
+
+		st.initialGas = st.msg.Gas()
+	*/
+	balance = vm.Sub(balance, vm.ValueFromUint256(mgval))
+	state.SetBalance(tx.Sender, balance)
+	return nil
+}
+
+func refundGas(tx vm.Transaction, gasLeft vm.Gas, state vm.WorldState) {
+
+	// Return wei for remaining gas, exchanged at the original rate.
+	refund := new(uint256.Int).Mul(new(uint256.Int).SetUint64(uint64(gasLeft)), tx.GasPrice.ToUint256())
+
+	cur := state.GetBalance(tx.Sender)
+	updated := new(uint256.Int).Add(cur.ToUint256(), refund)
+	state.SetBalance(tx.Sender, vm.ValueFromUint256(updated))
+
+	/*
+		// TODO: track block-wide gas usage
+		// Also return remaining gas to the block gas counter so it is
+		// available for the next transaction.
+		st.gp.AddGas(st.gas)
+	*/
+}
+
+// IntrinsicGas computes the 'intrinsic gas' for a message with the given data.
+func IntrinsicGas(transaction vm.Transaction) (vm.Gas, error) {
+	// Set the starting gas for the raw transaction
+	var gas uint64
+	if transaction.Recipient == nil {
+		gas = params.TxGasContractCreation
+	} else {
+		gas = params.TxGas
+	}
+	// Bump the required gas by the amount of transactional data
+	if len(transaction.Input) > 0 {
+		// Zero and non-zero bytes are priced differently
+		var nz uint64
+		for _, byt := range transaction.Input {
+			if byt != 0 {
+				nz++
+			}
+		}
+		// Make sure we don't exceed uint64 for all data combinations
+		if (math.MaxUint64-gas)/params.TxDataNonZeroGasEIP2028 < nz {
+			return 0, geth.ErrOutOfGas
+		}
+		gas += nz * params.TxDataNonZeroGasEIP2028
+
+		z := uint64(len(transaction.Input)) - nz
+		if (math.MaxUint64-gas)/params.TxDataZeroGas < z {
+			return 0, errGasUintOverflow
+		}
+		gas += z * params.TxDataZeroGas
+	}
+	accessList := transaction.AccessList
+	if accessList != nil {
+		gas += uint64(len(accessList)) * params.TxAccessListAddressGas
+		for _, tuple := range accessList {
+			gas += uint64(len(tuple.Keys)) * params.TxAccessListStorageKeyGas
+		}
+	}
+	return vm.Gas(gas), nil
+}
+
+func isInternal(transaction vm.Transaction) bool {
+	return transaction.Sender == vm.Address{}
+}
+
+// Source: https://github.com/Fantom-foundation/Sonic/blob/main/opera/contracts/evmwriter/evm_writer.go#L24
+
+// driverAddress is the NodeDriver contract address
+var driverAddress = common.HexToAddress("0xd100a01e00000000000000000000000000000000")
```

Comment:
maybe we should consider moving all these global variables and precompiled contracts to another file in the same package.